awx/main/tasks.py

```diff
@@ -602,25 +602,26 @@ def update_inventory_computed_fields(inventory_id, should_update_hosts=True):
 
 
 def update_smart_memberships_for_inventory(smart_inventory):
-    current = set(SmartInventoryMembership.objects.filter(inventory=smart_inventory).values_list('host_id', flat=True))
-    new = set(smart_inventory.hosts.values_list('id', flat=True))
-    additions = new - current
-    removals = current - new
-    if additions or removals:
-        with transaction.atomic():
-            if removals:
-                SmartInventoryMembership.objects.filter(inventory=smart_inventory, host_id__in=removals).delete()
-            if additions:
-                add_for_inventory = [
-                    SmartInventoryMembership(inventory_id=smart_inventory.id, host_id=host_id)
-                    for host_id in additions
-                ]
-                SmartInventoryMembership.objects.bulk_create(add_for_inventory)
-        logger.debug('Smart host membership cached for {}, {} additions, {} removals, {} total count.'.format(
-            smart_inventory.pk, len(additions), len(removals), len(new)
-        ))
-        return True  # changed
-    return False
+    with advisory_lock('update_smart_memberships_for_inventory-{}'.format(smart_inventory.id)):
```

Comment:
It somewhat bothers me to see this at a very different indent from `with transaction.atomic():`. For one (related to my own intent), this does actions to obtain the lock in no-op cases where it's not needed. Also, having both seems wrong to me. We could remove the transaction.atomic entirely and there should be no ill-effects because this is the _only_ code anywhere that ever edits the `SmartInventoryMembership` model.

If we kept the transaction.atomic, then honestly it seems like the most congruent fix would be to wrap it in a try-except block. Using the advisory lock is a very different kind of solution, so I'd prefer to just axe the atomic block for that case.

I should qualify that I still don't expect anything to break with your solution here, so there's no urgency to my comments.